detection_rules/version_lock.py

```diff
@@ -0,0 +1,164 @@
+                current_rule_lock.update(lock_info)
+
+                # add the min_stack_version to the lock if it's explicitly set
+                if rule.contents.metadata.min_stack_version is not None:
+                    current_rule_lock["min_stack_version"] = str(min_stack)
+
+            elif min_stack > latest_locked_stack_version:
+                # 3) on the latest stack, locking in a breaking change
+                previous_lock_info = {
+                    "rule_name": current_rule_lock["rule_name"],
+                    "sha256": current_rule_lock["sha256"],
+                    "version": current_rule_lock["version"],
+                }
+                current_rule_lock.setdefault("previous", {})
+
+                # move the current locked info into the previous section
+                current_rule_lock["previous"][str(latest_locked_stack_version)] = previous_lock_info
+
+                # overwrite the "latest" part of the lock at the top level
+                current_rule_lock.update(lock_info, min_stack_version=str(min_stack))
+
+            elif min_stack < latest_locked_stack_version:
+                # 4) on an old stack, after a breaking change has been made
+                assert str(min_stack) in current_rule_lock.get("previous", {}), \
+                    f"Expected {rule.id} @ v{min_stack} in the rule lock"
+
+                # TODO: Figure out whether we support locking old versions and if we want to
+                #       "leave room" by skipping versions when breaking changes are made.
+                #       We can still inspect the version lock manually after locks are made,
+                #       since it's a good summary of everything that happens
+                current_rule_lock["previous"][str(min_stack)] = lock_info
+                continue
+            else:
+                raise RuntimeError("Unreachable code")
+
+    for rule in rules:
+        if rule.id in newly_deprecated:
+            rule_deprecations[rule.id] = {
+                "rule_name": rule.name,
+                "stack_version": current_stack_version,
+                "deprecation_date": rule.contents.metadata.deprecation_date
+            }
+
+    if not save_changes:
+        echo('run `build-release --update-version-lock` to update version.lock.json and deprecated_rules.json')
+        return list(changed_rules), list(new_rules), list(newly_deprecated)
+
+    new_hash = dict_hash(current_versions)
+
+    if versions_hash != new_hash:
+        save_etc_dump(current_versions, ETC_VERSION_LOCK_FILE)
+        echo('Updated version.lock.json file')
+
+    if newly_deprecated:
+        save_etc_dump(rule_deprecations, ETC_DEPRECATED_RULES_FILE)
+        echo('Updated deprecated_rules.json file')
+
+    echo(f' - {len(changed_rules)} changed rules')
+    echo(f' - {len(new_rules)} new rules')
+    echo(f' - {len(newly_deprecated)} newly deprecated rules')
```

Comment:
can you move this above L135 so that it _always_ echos

Reply:
so... this isn't new code just moved

Reply:
True, but this change has been needed for a while and is just a small move

Reply:
the thing is though, this function is called every time a package is generated.
so that means to merely construct a package causes these prints. that's just confusing imho, so I think it is actually in the right place

Reply:
but it only prints if `verbose` flag is passed. That is the current quickest way to determine rule changes since last version lock. The alternative would be to duplicate this code into a separate function and share it between this and a dedicated stats printer, which doesn't make much sense.

Alternatively, if you still don't like it, we can throw this under L139:

```python
    if not save_changes:
        echo(f' - {len(changed_rules)} changed rules pending lock')
        echo(f' - {len(new_rules)} new rules pending lock')
        echo(f' - {len(newly_deprecated)} newly deprecated rules pending lock')
        echo('run `build-release --update-version-lock` to update version.lock.json and deprecated_rules.json')
        return list(changed_rules), list(new_rules), list(newly_deprecated)
```

Reply:
> but it only prints if verbose flag is passed.

right now this prints regardless of the verbose flag, every time the lock file is modified. i think the current behavior is right, no?

Reply:
oh i see. I can swap these to be `click.echo`, then win win?

Reply:
the problem that remains now is that it will only be reached if changes are saved, since it will return on L137

Reply:
so we want something like this?
```python
if save_changes or verbose:
    click.echo(...)
```

Reply:
thats one option